server/broadcast.go

```diff
@@ -217,10 +217,11 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) {
 		}
 		cond.L.Unlock()
 
-		// if !eth.VerifySig(transcoderAddress, crypto.Keccak256(segHashes...), res.Sig) { // need transcoder address here
-		// 	glog.Error("Sig check failed for segment ", seg.SeqNo)
-		// 	return
-		// }
+		ticketParams := cxn.sess.OrchestratorInfo.GetTicketParams()
+		if !pm.VerifySig(ethcommon.BytesToAddress(ticketParams.Recipient), crypto.Keccak256(segHashes...), res.Sig) {
```

Comment:
@j0sh @yondonfu I think we would all enjoy further refactoring around ETH stuff; basically breaking it down into: (a) generic ETH helpers, and (b) Livepeer-specific stuff.

I think that `VerifySig` could live in package (a), such that `pm` would depend on (a) as well.

wdyt?

Reply:
Where did you envision the generic ETH helpers would live @eladmallel ?

Reply:
@angyangie in a new package. I think it's best to have this conversation with a graph of the current dependencies, and from there try to create a new graph of what things should look like.

We've had to do some back bending due to cyclic import problems.

I don't know of any official tools for this, but bumped into [this one](https://github.com/davecheney/graphpkg) that might be useful.

@j0sh do you have a go-to tool for visualizaing dependencies in Go?

Reply:
That's a great idea @eladmallel. I've never used one of these packages, but seems like it could indeed help us in organizing our dependencies/imports

Reply:
We could split out the helpers. Note that I don't really have objections to the redundancy within PM since that's a bit of a special case. Ethereum-related code isn't the only area where we technically could refactor; we also have a number of helpers to generate random bytes, and so forth. Do we pull `pm.RandBytes` into a special non-Ethereum, non-Livepeer helper?

This line of thinking extends further anytime we might have a new module that might be a good candidate for open sourcing [1], especially if the helpers required by the new module are disjoint from others such as `pm`. Each new "standalone" module potentially means another attached helper module.

We should be mindful of taking DRY too far and splintering the codebase or making the project structure unnatural to satisfy a very specific constraint (eg, minimal dependencies in PM just so it may be generalized or open sourced one day [2]). We already have eth/helpers.go for ETH-specific helpers, `common` for more general helpers, and so forth. Then we'll have questions about which new code is Livepeer-specific and what "might" be useful outside the context of Livepeer. It's very easy to burn cycles on issues such as code organization when the project structure becomes complex.

I'm more inclined to accept small peccadilloes as-is rather than aiming for perfection, and refactoring as-we-go when the pieces naturally fall into place to enable that. In other words, reorganization and refactoring is a continuous process, rather than a single end goal.

[1]  Eg, we extract a 'transcoder management layer' into a general worker management mechanism incorporating reputation, redundancy, failover, etc. I have no intentions of doing this right now, but just putting this out by way of example.

[2] We probably should formalize this separation and extract out `pm` sooner rather than later, lest we lose sight of the original intention, and later developers start to introduce changes that increase the coupling with the Livepeer code.

Reply:
@j0sh regarding [2], do you think it would suffice to leave the `pm` package and write a separate README indicating that it can be standalone package (despite being nested in go-livepeer) and highlight that Livepeer specific dependencies should not be introduced in the package?

Reply:
@j0sh I completely agree with your iterative approach, and that we should not strive for "perfect DRY".

Having said that, I also think that in the upcoming refactoring/cleanup session, it could be beneficial to re-org some of our packages.

From your note I'm not sure if I understand your POV for the very near term: do you think we should leave the current architecture as-is, or are there refactoring opportunities that you're in favor of? (e.g. splitting out DB from `common`? moving `pm` out, maybe a-la `lpms`?)

Reply:
> do you think it would suffice to leave the pm package and write a separate README indicating that it can be standalone package

We can certainly try to enforce it socially for now. Long-term, splitting it into a standalone repo also means the split is enforced by the compiler, so people would have to 'think before they push'. It would also be more beneficial as far as encouraging visibility, adoption and community participation around probabilistic micropayments as a concept.

> are there refactoring opportunities that you're in favor of? (e.g. splitting out DB from common? moving pm out, maybe a-la lpms?)

Those two sound good, yes.

If we take the view that the `common` package is actually for *helpers* rather than common code that's used by most modules, then the DB does feel out of place there.

We'll might still end up with issues if we're not consistent with how the DB is used. Thus far, all code has used the DB directly because there hasn't been a need to abstract it out. PM inverts the flow of control by making the DB implement a PM-specific storage interface. The DB then becomes pm-aware, and it's not really possible to export data or features that are specific to the storage layer, at least not without yet another central module for `types` or so forth (the `net` module in goclient has sort of taken on this role elsewhere). We also end up with more glue code to make modules interoperate (eg, livepeer.go).

Anyway, I'm really OK with the split, just highlighting some potential issues that we should be aware of wrt the inversion of control.

Reply:
Seems like we can continue discussion here: https://github.com/livepeer/go-livepeer/issues/717